osbuild/sources.py

```diff
@@ -1,14 +1,76 @@
+import concurrent.futures
 import abc
 import contextlib
 import os
 import json
 import tempfile
+from typing import List, Dict
+from enum import Enum
 
+from osbuild.formats_common import OSBuildError
 from . import host
 from .objectstore import ObjectStore
 from .util.types import PathLike
 
 
+class SourceErrorKind(Enum):
+    UNKNOWN = 0
+    CHECKSUM = 1
+    # OSTree errors
+    OSTREE_INIT = 2
+    OSTREE_CONFIG = 3
+    OSTREE_REMOTE_ADD = 4
+    OSTREE_GPG_IMPORT = 5
+    OSTREE_PULL = 6
+    OSTREE_REMOTE_DELETE = 7
+    # Skopeo errors
+    SKOPEO_COPY = 8
+    SKOPEO_INSPECT = 9
+    # Curl errors
+    CURL_BAD_INPUT = 10
+    CURL_INTERNAL_ERROR = 11
+    CURL_NETWORKING = 12
+    CURL_CREDENTIALS = 13
+    CURL_STORAGE = 14
+    CURL_SECURITY = 15
```

Comment:
Having per-source specific code kinda defeats the point. The idea would be to define some generic source related error codes and in the specific source we map the error that we got to any of the pre-defined ones.
Which specific ones to include is the very tricky part and we should start we fewer and extend them. They need to be informative and ideally actionable. Like, if we cannot download because the file is not there, or network error, invalid credentials or the file could be downloaded but was corrupted. Or more generically: user error, network error, content error, and storage/worker error? I.e. Composer could do something with the last info, e.g. stop the worker, clean the disk space and restart the worker. But there is not much that Composer or a user can do with `SKOPEO_INSPECT`. 

NB: one of osbuild's guiding principals is that the core of osbuild does not "know" with the specifics of the image building process. Hence everything is abstracted in sources, inputs, mounts, devices and stages. There should never be a mention of any specific instance of any of those.